templates/about.html

```diff
@@ -1,6 +1,6 @@
 {% extends "layout.html" %}
 {% block styles %}
-<link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='stylesheet.css') }}" />
```

Comment:
I disagree with hardcoding the path to the `static` folder. Flask allows you to configure where the folder should be located and how it should be called, so hardcoding to `/static/` isn't ideal. Using `url_for` deals with cases where the folder doesn't have the default name or location.

See: https://flask.palletsprojects.com/en/1.1.x/api/?highlight=flask#flask.Flask

Reply:
Flask allows to configure static folder for setups where flask **is** the web server. But in our case flask is only the backend, so I'd doesn't own and doesn't serve static.
The url to the static is hard-coded in apache config, in most of the frameworks' css and in templates themselves.
The usage of url_for in this case added nothing to flexibility, but added load to the backend, which is our bottleneck.

_url_for is mostly intended for building complex hierarchies of **endpoints** with blueprints and such, allowing it to be maintainable_

Reply:
And when we do use `url_for`, it's better to call it from flask code, passing the result to templates, rather than call it from templates directly. This separates logic from representation and allows easy refactoring, as it's limited to the python code only.

Reply:
Doesnt Flask generate the html and thus img src tags and alike?

In order to move to a cookieless seperate static content domain we should have the ability to change the url of content in templates (if possible from a central point).
Apache webserver still serves static content directly in any case.

Reply:
Our scripts and styles are compiled and outputted to `app.static_folder`. We haven't changed it, but hardcoding to `/static` just makes it difficult if we ever decide to change the location of the folder.

https://github.com/KSP-SpaceDock/SpaceDock/blob/master/app.py#L35



> The usage of `url_for` in this case added nothing to flexibility, but added load to the backend, which is our bottleneck.

See above, we're compiling our assets and outputting them to `app.static_folder`. I would agree with you, but it's not what it currently looks like.




> And when we do use `url_for`, it's better to call it from flask code, passing the result to templates, rather than call it from templates directly.

I think the way it is right now is exactly the former. If I understood it correctly, there wouldn't be a difference if we used a jinja variable for the static folder or if we kept using `url_for`.

> Flask allows to configure static folder for setups where flask **is** the web server. But in our case flask is only the backend, so I'd doesn't own and doesn't serve static.

From my understanding of past and current SpaceDock/KerbalStuff code, I think we're actually using Flask as a web server, otherwise nothing we did would be possible (rendering, templating, etc). From what I understood from the previous version that was maintained by SirCmpwn, you'd need a reverse proxy to serve the content from KerbalStuff, and I think it's exactly what Apache is doing.

Reply:
Currently only urls for endpoints are generated by flask in templates, which is proper, as they are owned by it. All statics are hard-coded, but, I repeat, they're also hard-coded in many frameworks we use; so there's not many options as to where the .css, fonts and .js could be. And static images are hard-coded into .sass styles.
So as it is, there's no single point where statics are defined, and I don't see a way to create one. Certainly not with flask.

Reply:
Also, at this point I'm more concerned with backend performance and stability than with cookieless statics.

Reply:
> I'm more concerned with backend performance and stability

I think we should benchmark it to confirm if there's a significant gain from hardcoding static paths.

Reply:
> Our scripts and styles are compiled and outputted to app.static_folder. We haven't changed it, but hardcoding to /static just makes it difficult if we ever decide to change the location of the folder.
> https://github.com/KSP-SpaceDock/SpaceDock/blob/master/app.py#L35

This doesn't change the fact that the images are hardcoded into these same styles. And no matter where you put _them_, they're still served at an url configured in apache.

Also, one of the things I want to change (actually, already changed in my local branch) is to switch to compilation of coffee and scss by dedicated node tools instead of the poorly-supported third-party python ports of them. Aside from transpilation of the result javascript this will allow us to switch from serving multiple huge `some_framework.min.js` to our own `main.bundle.js` using webpack.

> there wouldn't be a difference if we used a jinja variable for the static folder

There would be a difference in performance, as the current static url could be cached once. And it would be a world of difference when you start refactoring the code. Not even PyCharm is able to refactor python code in jinja templates.

> From my understanding of past and current SpaceDock/KerbalStuff code, I think we're actually using Flask as a web server, otherwise nothing we did would be possible (rendering, templating, etc).

Currently KerbalStuff code is run by the gunicorn WSGI engine, only serving what is passed to it by apache via proxy pass, namely, all the non /static requests. If some in response to some of these requests the API generates rendered HTML it doesn't make it into a web server.

Reply:
> I think we should benchmark it to confirm if there's a significant gain from hardcoding static paths.

Does any of us know how to perform proper high-load testing? I don't. What I do know is that not every bad design choice needs to be benchmarked in order to decide that it needs fixing. And I also know from experience that mixing responsibilities and misusing flask and jinja leads to poorly performing web applications with hard-to-diagnose problems.

It's much simpler not to make mistakes than to track bugs and performance issues afterward.